### ISSUE TITLE
WarpSync reentrancy bug for minting REP.

### ISSUE THREAD

**Issue by MicahZoltu** [AugurProject/augur#5122]
TL;DR: `mint` calls into user code and thus you can execute a reentrancy attack against WarpSync's `awardRep` function, with an added bonus of `createMarket`'s RepBond minting.

1. `Market.finalize` on the WarpSync market.
1. `WarpSync.notifyMarketFinalized` will set `winningPayoutDistributionHash` and then call into `WarpSync.notifyMarketFinalized`
1. Before updating any state, that will call into `WarpSync.recordMarketFinalized`
1. Before updating any state, that will call into `WarpSync.awardRep`
1. Before updating any state, that will call into `ReputationToken.mintForWarpSync`
1. Before updating any state, that will call into `ReputationToken.mint`
1. Before updating any state, that will call into `ReputationToken._mint`
1. `ReputationToken.supply` will be increased.
1. `ReputationToken.balances[tx.origin]` will be incresed.
1. `ReputationToken._callTokensReceived` will be called.
1. --> If `tx.origin` has registered an ERC1820 contract, it will be called.  This gives the caller an opportunity to call `WarpSync.notifyMarketFinalized()` again.
1. During unwind, `ReputationToken._callTokensReceived` doesn't do anything else.
1. During unwind, `ReputationToken._mint` only fires events.
1. During unwind, `ReputationToken.mint` only fires events.
1. During unwind, `ReputationToken.mintForWarpSync` will call `Universe.updateForkValues`.
1. `Universe.updateForkValues` only executes assignments, it doesn't do any checks other than `isForking`.
1. During unwind, `WarpSync.awardRep` doesn't do anything else.
1. During unwind, `WarpSync.recordMarketFinalized` will only execute idempotent assignments.
1. During unwind, `WarpSync.notifyMarketFinalized` will call `WarpSync.createMarket`.
1. --> `WarpSync.createMarket` will finally make the update necessary to prevent this method from being called again (but too late for reentrancy protection).
1. `WarpSync.createMarket` only does idempotent assignments with the exception of another call to `ReputationToken.mintForWarpSync`, which will mint excessive bonds for the next market (amplifying the attack if the attacker can repeat multiple days in a row).

**Comment by MicahZoltu** [comment]
Gah, nevermind.  Missed a check.  *grumble*

For the historians: the check is that at the start of `Market.finalize` `isFinalized` is checked which looks at `winningPayoutDistributionHash`.  `winningPayoutDistributionHash` is then set prior to calling into `warpSync.notifyMarketFinalized()`.

**Comment by MicahZoltu** [comment]
https://github.com/AugurProject/augur/issues/5123 is still possible, but not as exploitable as this one.
